// express-validator/filter/dist/sanitize.dev.js
"use strict";

var validator = require('validator');

var _require = require('../utils/filters'),
    isSanitizer = _require.isSanitizer;

var selectFields = require('../utils/select-fields');

module.exports = function (fields, locations) {
  var sanitizers = [];
  fields = Array.isArray(fields) ? fields : [fields];

  var middleware = function middleware(req, res, next) {
    selectFields(req, {
      fields: fields,
      locations: locations,
      sanitizers: sanitizers
    });
    next();
  };

  Object.keys(validator).filter(isSanitizer).forEach(function (methodName) {
    var sanitizerFn = validator[methodName];

    middleware[methodName] = function () {
      for (var _len = arguments.length, options = new Array(_len), _key = 0; _key < _len; _key++) {
        options[_key] = arguments[_key];
      }

      sanitizers.push({
        sanitizer: sanitizerFn,
        options: options
      });
      return middleware;
    };
  });

  middleware.customSanitizer = function (sanitizer) {
    sanitizers.push({
      sanitizer: sanitizer,
      custom: true,
      options: []
    });
    return middleware;
  };

  return middleware;
};
